### PR TITLE
CI: Create AppImage artifacts

### PR DESCRIPTION
**Pull request by qurious-pixel**
Create AppImage and binary artifacts available for download from Github Actions

Make AppImage executable with the following command
`chmod +x Dolphin-*.AppImage`

Launch Dolphin
`./Dolphin-*.AppImage`

### CODE REVIEW

**Comment by stenzek** [comment]
Using AppImages will create issues when we eventually merge wayland support (I still need to finish it). The deploy plugin does not support wayland, and the authors refuse to support it. Same thing with Qt 6.

(also, if this were to go ahead, we'd probably want to plug it into Dolphin's existing CI infrastructure, rather than using actions)

**Comment by qurious-pixel** [comment]
> Using AppImages will create issues when we eventually merge wayland support (I still need to finish it). The deploy plugin does not support wayland, and the authors refuse to support it. Same thing with Qt 6.
> 

I am not sure what you mean. Wayland has been supported in AppImages since 2018. Do you mean QT6?
<details>
<summary> Fedora34 in Wayland </summary>

DolphinWaylandAppimage


</details> 

> (also, if this were to go ahead, we'd probably want to plug it into Dolphin's existing CI infrastructure, rather than using actions)

I looked at the Dolphin CI, but did not see it produce any artifacts. If there is a way to incorporate this with one of the workers to produce artifacts, I am willing to contribute.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          chmod a+x ./squashfs-root/runtime
+          /tmp/squashfs-root/AppRun $GITHUB_WORKSPACE/squashfs-root/usr/bin/dolphin-emu -appimage -unsupported-allow-new-glibc -no-copy-copyright-files -no-translations -bundle-non-qt-libs
+          ls . | grep *.AppImage
+          mv $(ls . | grep *.AppImage) $GITHUB_WORKSPACE/artifacts
+          ls -al $GITHUB_WORKSPACE/artifacts
+        continue-on-error: true  
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: Dolphin_Emulator-x86_64-Linux
```
Comment (Margen67) [inline .github/workflows/build.yml:94]:
```suggestion
          name: Dolphin_Emulator-x86_64-${{ runner.os }}
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+          path: artifacts/
```
Comment (Margen67) [inline .github/workflows/build.yml:95]:
```suggestion
          path: artifacts/
          if-no-files-found: error
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+          sudo add-apt-repository -y ppa:beineri/opt-qt-5.15.2-focal
+          sudo apt-get install -y dialog libbluetooth-dev liblz4-dev libopus-dev libsystemd-dev \
+          libzip-dev libzstd-dev libasound2-dev libpulse-dev pulseaudio libsfml-dev libminiupnpc-dev libmbedtls-dev libpugixml-dev \
+          liblzo2-dev qt515base qt5-qmake libxi-dev libavcodec-dev libudev-dev libusb-1.0-0-dev libevdev-dev libhidapi-dev libavformat-dev libavdevice-dev \
+          libfmt-dev libwayland-dev libxrandr-dev libglu1-mesa-dev libcurl4-openssl-dev zenity gettext ccache cmake ninja-build \
+      - name: Install Packages if failed
+        id: packages2
+        if: steps.packages.outcome=='failure'
+        run: |
+          sudo add-apt-repository -y ppa:beineri/opt-qt-5.15.2-focal
```
Comment (Margen67) [inline .github/workflows/build.yml:43]: Unless the PPA doesn't persist somehow this should only be done once.
```suggestion
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+          mkdir -p squashfs-root/usr/share/icons && cp ./squashfs-root/dolphin-emu.svg ./squashfs-root/usr/share/icons
+          mkdir -p squashfs-root/usr/share/icons/hicolor/scalable/apps && cp ./squashfs-root/dolphin-emu.svg ./squashfs-root/usr/share/icons/hicolor/scalable/apps
+          mkdir -p squashfs-root/usr/share/pixmaps && cp ./squashfs-root/dolphin-emu.svg ./squashfs-root/usr/share/pixmaps
+          mkdir -p squashfs-root/usr/share/dolphin-emu
+          cp -R Data/Sys/Themes ./squashfs-root/usr/share/dolphin-emu
+          chmod a+x ./squashfs-root/AppRun
+          chmod a+x ./squashfs-root/runtime
+          /tmp/squashfs-root/AppRun $GITHUB_WORKSPACE/squashfs-root/usr/bin/dolphin-emu -appimage -unsupported-allow-new-glibc -no-copy-copyright-files -no-translations -bundle-non-qt-libs
+          ls . | grep *.AppImage
+          mv $(ls . | grep *.AppImage) $GITHUB_WORKSPACE/artifacts
```
Comment (Margen67) [inline .github/workflows/build.yml:88]: Why not just do this?
```suggestion
          ls *.AppImage
          mv *.AppImage $GITHUB_WORKSPACE/artifacts
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          libzip-dev libzstd-dev libasound2-dev libpulse-dev pulseaudio libsfml-dev libminiupnpc-dev libmbedtls-dev libpugixml-dev \
+          liblzo2-dev qt515base qt5-qmake libxi-dev libavcodec-dev libudev-dev libusb-1.0-0-dev libevdev-dev libhidapi-dev libavformat-dev libavdevice-dev \
+          libfmt-dev libwayland-dev libxrandr-dev libglu1-mesa-dev libcurl4-openssl-dev zenity gettext ccache cmake ninja-build \
+      - name: Install Packages if failed
+        id: packages2
+        if: steps.packages.outcome=='failure'
+        run: |
+          sudo add-apt-repository -y ppa:beineri/opt-qt-5.15.2-focal
+          sudo apt-get install -y build-essential dialog dpkg libbluetooth-dev liblz4-dev liblzma-dev libssl-dev libopus-dev libpng-dev libsystemd-dev \
+          libzip-dev libzstd-dev zlib1g-dev libasound2-dev libpulse-dev pulseaudio p7zip p7zip-full libsfml-dev libminiupnpc-dev libmbedtls-dev libpugixml-dev \
```
Comment (Margen67) [inline .github/workflows/build.yml:45]: `p7zip` is a dependency of `p7zip-full` (there are probably other redundant packages but I cba to go through them)
```suggestion
          libzip-dev libzstd-dev zlib1g-dev libasound2-dev libpulse-dev pulseaudio p7zip-full libsfml-dev libminiupnpc-dev libmbedtls-dev libpugixml-dev \
```

**Comment by sykhro** [comment]
> Using AppImages will create issues when we eventually merge wayland support (I still need to finish it). The deploy plugin does not support wayland, and the authors refuse to support it. Same thing with Qt 6.
> 
> (also, if this were to go ahead, we'd probably want to plug it into Dolphin's existing CI infrastructure, rather than using actions)

I have been able to create Wayland AppImages by including the `platforms/, wayland-shell-integration/, wayland-decoration-client/` folders from Qt.

**Comment by gsantner** [comment]
Looking forward to this. I use Debian 11 on multiple hosts and Debian dropped `dolphin-emu` as it couldn't deliver the required quality -> `https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=976530`.  
Currently using this one build here `https://www.appimagehub.com/p/1567651`, but really waiting for official & regular builds.  
I don't really care about dolphin on wayland, a first official release without it would be totally acceptable.  

(No, don't want to use flatpak/snap for dolphin)

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,95 @@
+name: Dolphin-AppImage
+
+on: 
+  push:
+
+jobs:
+  dolphin-appimage:
+    name: dolphin bundled as appimage
+    runs-on: ubuntu-latest
```
Comment (gsantner) [inline .github/workflows/build.yml:9]: If you use ubuntu-latest (and so a quite recent libc), it basically defeats the purpose of appimage (run on any "as old as possible OS"). Suggestion: 16.04, 18.04 or Debian 9.

**Comment by gsantner** [comment]
Looking forward to this. I use Debian 11 on multiple hosts and Debian dropped `dolphin-emu` as it couldn't deliver the required quality -> `https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=976530`.  
Currently using this one build here `https://www.appimagehub.com/p/1567651`, but really waiting for official & regular builds.  
I don't really care about dolphin on wayland, a first official release without it would be totally acceptable.  

(No, don't want to use flatpak/snap for dolphin)

Any chance this gets merged once it produces the appimage artifact?

File: .github/workflows/build.yml (hunk 6 above)
Comment (qurious-pixel) [reply .github/workflows/build.yml:9]: Yes, you are correct. On my version I run ubuntu-18.04 ( LTS ).

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,86 @@
+name: Dolphin-AppImage
+
+on: 
+  push:
```
Comment (Margen67) [inline .github/workflows/build.yml:4]:
```suggestion
on: push
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      - name: Package Dolphin-AppImage
+        env:
+          QT_BASE_DIR: /opt/qt515
+          QTDIR: $QT_BASE_DIR
+        run: |
+          export PATH=$QT_BASE_DIR/bin:$PATH
+          echo "PATH $PATH"
+          qmake -v
+          cd /tmp 
+          curl -sLO "https://github.com/probonopd/linuxdeployqt/releases/download/continuous/linuxdeployqt-continuous-x86_64.AppImage"
```
Comment (Margen67) [inline .github/workflows/build.yml:59]: Using only `s` suppresses errors.
```suggestion
          curl -sSLO "https://github.com/probonopd/linuxdeployqt/releases/download/continuous/linuxdeployqt-continuous-x86_64.AppImage"
```

File: .github/workflows/build.yml (hunk 7 above)
Comment (gsantner) [reply .github/workflows/build.yml:4]: @qurious-pixel you can just press OK here and github automatically applies the suggested changes

**Comment by delroth** [comment]
I only noticed this PR just now:

- Dolphin does not officially support Flatpak or AppImages or any similar containerized tech. Use your distro provided packages, and use a better distro if that's not an option (~all distros package our stable releases, many distro package our beta builds incl. Arch, Gentoo, NixOS, and more).
- We are not using GitHub Workflows and have no plans to do so.

**Comment by kenderipa** [comment]
> * Dolphin does not officially support Flatpak or AppImages or any similar containerized tech.

Why? It allows a far more clean system not burdened by hundreds of packages since all dependecies are included in the AppImasge, it's portable, easy to use, easy to upgrade and easy to integrate in the sysyem (with an AppIamage manager or daemon). Also, like it or not, it's the future Linux software is moving right now, and this future looks pretty bright to me.

**Pull request closed without merging**